app/services/fact_importer.rb

```diff
@@ -66,11 +68,17 @@ def add_new_facts
     facts_to_create = facts.keys - db_facts.keys
     # if the host does not exists yet, we don't have an host_id to use the fact_values table.
     if facts_to_create.present?
-      method          = host.new_record? ? :build : :create!
-      fact_names      = fact_name_class.group(:name).maximum(:id)
+      method = host.new_record? ? :build : :create!
+      # :type is needed because custom facts usually inherits from FactName so they would be included in the list
+      fact_names = fact_name_class.where(:type => fact_name_class).group(:name).maximum(:id)
       facts_to_create.each do |name|
-        host.fact_values.send(method, :value => facts[name],
-                              :fact_name_id  => fact_names[name] || fact_name_class.create!(:name => name).id)
+        begin
+          host.fact_values.send(method, :value => facts[name],
+                                :fact_name_id  => fact_names[name] || fact_name_class.create!(:name => name).id)
```

Comment:
The reason was that if we used `FactName.group(:name).maximum(:id)` it loaded both facts with the same name but different type because the other type inherited from FactName (STI). Then the other fact was used in `maximum`, because it was created later it had higher id. And because `fact_names[name]` existed it tried to create FactValue with this existing fact name of different type which effectively broke uniqueness validation.

Example: _timestamp value of type CustomFactName exists but value _timestamp of type FactName does not exist and new import of FactName facts includes this _timestamp fact. What happens - `facts_to_create` contains _timestamp, because there's no such FactName record but in fact_names we got id of CustomFactName, therefore we try to `create!` new FactValue with fact_name_id set to CustomFactName record with different value, so it's FactValue uniqueness check that fails, we can't have to values for same fact per host. Simply said the issue is that we tried to create value for different fact type.